### ISSUE TITLE
User reported issue: short circuit created with switch open

### ISSUE THREAD

**Issue by oliver-phet** [phetsims/circuit-construction-kit-common#693]
A user wrote in with this bug and I was able to quickly reproduce:

> **Test device**
> Dell Latitude E6530
> Dell Chromebook P26T
> 
> **Operating System**
> Windows 7 
> Chrome OS
> 
> **Browser**
> Chrome
> 
> **Problem description**
> A certain configuration leads to an inaccurate short circuit when a switch is opened.
> 
> **Steps to reproduce**
> In intro mode create a circuit in this configuration:
> Positive terminal of battery ->wire ->switch ->light bulb ->wire ->eraser ->wire ->negative terminal of battery
> 
> With the switch closed, nothing happens (as expected). When the switch is opened the bulb lights up and the battery and eraser catch fire. This should not happen since no current should move as the circuit is open. 
> 
> The same thing happens when a dollar bill is used.
> 
> If the current is investigated with the Ammeter it shows differing current throughout which also shouldn't happen.
> 
> **Visuals**
> ![Circuit Construction Kit Error Open](https://user-images.githubusercontent.com/5090275/116472331-02e08880-a833-11eb-885a-90c1c60720f2.png)
>

**Comment by oliver-phet** [comment]
In some quick testing on my end, it appears to possibly be related to the eraser.

**Comment by arouinfar** [comment]
I built a few different similar circuits (including the one below), but couldn't reproduce the issue.
<img src="https://user-images.githubusercontent.com/8419308/116477670-2955f200-a83a-11eb-97cf-8a5556f9fbd6.png" width=250>


The screenshot from the user is exhibiting the bunched-up current that was also reported in https://github.com/phetsims/circuit-construction-kit-common/issues/281 which makes me think it's somehow related.

Assigning to @samreid to take a look.

**Comment by oliver-phet** [comment]
Oh weird - I was immediately able to reproduce, but now when I try, it seems more elusive.

Here's a circuit I just built that displays the bug with no electron bunching:
<img width="1090" alt="image" src="https://user-images.githubusercontent.com/5090275/116479680-4a6c1200-a83d-11eb-82a2-a586f8a0e0c0.png">

**Comment by oliver-phet** [comment]
Actually, there does seem to be some bunching. Here's that same circuit I built:
![Kapture 2021-04-28 at 16 19 59](https://user-images.githubusercontent.com/5090275/116479989-c8301d80-a83d-11eb-9780-ff431d7bf48a.gif)

You can see the current arrows and electrons move through the bulb, but then seem to disappear/stack up.

**Comment by oliver-phet** [comment]
And to be clear - there are no "0 length" wires or overlapped contacts in my circuit.

**Comment by arouinfar** [comment]
So strange! Thanks for investigating @oliver-phet. The electons are missing from the bulb in https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-828817870 so there might be some bunching going on there too.

**Comment by samreid** [comment]
Some observations:

* I was able to reproduce the problem 3-4 times.
* I reproduced the problem with a dollar bill in place of the eraser (noting they have the same resistance of 1000000000)
* Then I went through a phase where, following the same procedure as above, I was not able to reproduce the problem about 5 times in a row.
* Next, I saw a different symptom, where an eraser in an open circuit flickered on fire, but only while dragging (other symptoms were not present).

I don't know what is causing this, but hypothesized that it is a numerical error with `1000000000` resistance.  I tried lowering it by a factor of 10 and saw the same problem, then by a factor of 100 and saw the same symptom with lower current.  But I identified a better way to model true insulators--we can omit them from the circuit solution as we do for open switches.  I don't recall if we wanted erasers and dollars to be pure insulators or just very high resistors.  But I'll try modeling them as pure insulators to see what happens.  But I'm skeptical I will be able to test this accurately, since there are some times where I test 5 times in a row (with no code changes) and cannot see the problem.

**Comment by samreid** [comment]
@arouinfar can you please review my notes & the commit, and advise?

**Comment by samreid** [comment]
@arouinfar pointed out that it reads 0 Ohms, that should be corrected.

**Comment by samreid** [comment]
We would like to revert the prior commit, since we would like the resistance to be non-infinite.

**Comment by samreid** [comment]
After that, we will ask @KatieWoe and QA to investigate this issue, see if it is possible to reliably reproduce.

@kathy-phet reminds @samreid to put it in the issues to test/verify column on the QA project board.

**Comment by KatieWoe** [comment]
This seems to happen more noticeably with higher voltages, and doesn't seem to need a 0 length wire, though it does make it easier.
<img width="954" alt="notshorthascurrent" src="https://user-images.githubusercontent.com/41024075/127910984-eb086263-e42e-418b-811d-0888cafb44ce.png">
![nozerobug](https://user-images.githubusercontent.com/41024075/127910986-35f57da2-ba27-4987-ad5f-1f0a4ecbcb5f.gif)
![zerolengthbug1](https://user-images.githubusercontent.com/41024075/127910990-803cdc80-d001-4a10-925f-c86dbdd702c4.gif)
![zerolengthhighvolt](https://user-images.githubusercontent.com/41024075/127910992-fc47ff75-ff97-47ec-b40a-1270ebf3b3bf.gif)

**Comment by KatieWoe** [comment]
Clarification: The above was in published

**Comment by KatieWoe** [comment]
I can have trouble reproducing this with non-lightbulbs, but it still happens in master. Turning up voltage helps.
<img width="959" alt="stillmaster" src="https://user-images.githubusercontent.com/41024075/128042133-b4cf46af-c07d-42f5-88f1-140e51972399.png">
![probleminmaster](https://user-images.githubusercontent.com/41024075/128042135-8ea214ca-2d06-47ca-8bf1-c0a5bb823a25.gif)

**Comment by samreid** [comment]
I reverted the change that omits eraser and dollar from the circuit matrix.

**Comment by samreid** [comment]
After some iteration, I was able to get to this error case.  Note the current changes as you adjust the wire length:

![Kapture 2021-08-12 at 14 32 28](https://user-images.githubusercontent.com/679486/129265678-c370668b-c17f-47b4-919a-352924fc5894.gif)

**Comment by samreid** [comment]
After creating the above movie, I tried removing a few components to simplify the circuit.  I was able to remove several parts and keep the buggy behavior, but then after removing something else (a wire?) it could no longer produce the problem at all, even when adding everything back.

**Comment by samreid** [comment]
Here's a case where attaching the light bulb triggers the problem:

![Kapture 2021-08-12 at 14 59 44](https://user-images.githubusercontent.com/679486/129269123-eaaf798e-4bdf-4107-8025-ace7d4a90625.gif)

**Comment by samreid** [comment]
For this circuit, here are the equations:

![image](https://user-images.githubusercontent.com/679486/129291785-21d7cce9-0f2a-440d-b467-7b4abd8809c1.png)

```
Debugging circuit: 
resistors:node1 -> node2 @ 2.07037670465044e-10 Ohms,node2 -> node3 @ 10 Ohms,node4 -> node1 @ 0.000001 Ohms, 
batteries: node0 -> node4 @ 100000 Volts, 
currentSources: 
```

```
V0=0
1000000*V1+-1000000*V4+-4830038889.801163*V2+4830038889.801163*V1=0
4830038889.801163*V2+-4830038889.801163*V1+-0.1*V3+0.1*V2=0
0.1*V3+-0.1*V2=0
-I0_4+-1000000*V1+1000000*V4=0
-V0+V4=100000
```

```
A=
dim: 6x6
0 1 0 0 0 0 
0 0 4831038889.801163 -4830038889.801163 0 -1000000 
0 0 -4830038889.801163 4830038889.901163 -0.1 0 
0 0 0 -0.1 0.1 0 
-1 0 -1000000 0 0 1000000 
0 -1 0 0 0 1 
```

```
z=
dim: 6x1
0 
0 
0 
0 
0 
100000 
```

```
unknowns=
I0_4
V0
V1
V2
V3
V4
```

```
x=
dim: 6x1
0.02496337890625 
0 
99999.99999997504 
99999.99999997503 
99999.99999997504 
100000 
```

**Comment by samreid** [comment]
Using bignumber.js v9.0.1, I adapted LU Decomposition to be arbitrary-precision, then with this patch:

<details>

```js
Index: js/LUDecomposition.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/LUDecomposition.js b/js/LUDecomposition.js
--- a/js/LUDecomposition.js	(revision c63a40f4ec3f0c86260c4fe18a7708079925afe5)
+++ b/js/LUDecomposition.js	(date 1628857320958)
@@ -11,6 +11,8 @@
 
 const ArrayType = window.Float64Array || Array;
 
+BigNumber.config({ DECIMAL_PLACES: 1000 })
+
 class LUDecomposition {
   constructor( matrix ) {
     let i;
@@ -20,7 +22,12 @@
     this.matrix = matrix;
 
     // TODO: size!
-    this.LU = matrix.getArrayCopy();
+    // this.LU = matrix.entries.map(entry=>new BigNumber(entry))
+    this.LU = [];
+    matrix.entries.forEach(entry=>{
+      this.LU.push(new BigNumber(entry))
+    });
+
     const LU = this.LU;
     this.m = matrix.getRowDimension();
     const m = this.m;
@@ -31,7 +38,7 @@
       this.piv[ i ] = i;
     }
     this.pivsign = 1;
-    const LUcolj = new ArrayType( m );
+    const LUcolj = new Array(m);
 
     // Outer loop.
 
@@ -39,7 +46,7 @@
 
       // Make a copy of the j-th column to localize references.
       for ( i = 0; i < m; i++ ) {
-        LUcolj[ i ] = LU[ matrix.index( i, j ) ];
+        LUcolj[ i ] = new BigNumber(LU[ matrix.index( i, j ) ]);
       }
 
       // Apply previous transformations.
@@ -47,13 +54,16 @@
       for ( i = 0; i < m; i++ ) {
         // Most of the time is spent in the following dot product.
         const kmax = Math.min( i, j );
-        let s = 0.0;
+        let s =new BigNumber(0);
         for ( k = 0; k < kmax; k++ ) {
           const ik = matrix.index( i, k );
-          s += LU[ ik ] * LUcolj[ k ];
+          const a = new BigNumber(LU[ik]);
+          const b = LUcolj[k];
+          s = s.plus(a.times(b));
+          // s += LU[ ik ] * LUcolj[ k ];
         }
 
-        LUcolj[ i ] -= s;
+        LUcolj[ i ] = LUcolj[ i ].minus(s)
         LU[ matrix.index( i, j ) ] = LUcolj[ i ];
       }
 
@@ -61,7 +71,7 @@
 
       let p = j;
       for ( i = j + 1; i < m; i++ ) {
-        if ( Math.abs( LUcolj[ i ] ) > Math.abs( LUcolj[ p ] ) ) {
+        if (  LUcolj[ i ].abs()  > LUcolj[ p ].abs() ) {
           p = i;
         }
       }
@@ -83,7 +93,7 @@
 
       if ( j < m && LU[ this.matrix.index( j, j ) ] !== 0.0 ) {
         for ( i = j + 1; i < m; i++ ) {
-          LU[ matrix.index( i, j ) ] /= LU[ matrix.index( j, j ) ];
+          LU[ matrix.index( i, j ) ] = LU[ matrix.index( i, j ) ].dividedBy(LU[ matrix.index( j, j ) ]);
         }
       }
     }
@@ -97,7 +107,7 @@
   isNonsingular() {
     for ( let j = 0; j < this.n; j++ ) {
       const index = this.matrix.index( j, j );
-      if ( this.LU[ index ] === 0 ) {
+      if ( this.LU[ index ].isZero() ) {
         return false;
       }
     }
@@ -108,44 +118,44 @@
    * @public
    *
    * @returns {Matrix}
-   */
-  getL() {
-    const result = new Matrix( this.m, this.n );
-    for ( let i = 0; i < this.m; i++ ) {
-      for ( let j = 0; j < this.n; j++ ) {
-        if ( i > j ) {
-          result.entries[ result.index( i, j ) ] = this.LU[ this.matrix.index( i, j ) ];
-        }
-        else if ( i === j ) {
-          result.entries[ result.index( i, j ) ] = 1.0;
-        }
-        else {
-          result.entries[ result.index( i, j ) ] = 0.0;
-        }
-      }
-    }
-    return result;
-  }
-
-  /**
-   * @public
-   *
-   * @returns {Matrix}
-   */
-  getU() {
-    const result = new Matrix( this.n, this.n );
-    for ( let i = 0; i < this.n; i++ ) {
-      for ( let j = 0; j < this.n; j++ ) {
-        if ( i <= j ) {
-          result.entries[ result.index( i, j ) ] = this.LU[ this.matrix.index( i, j ) ];
-        }
-        else {
-          result.entries[ result.index( i, j ) ] = 0.0;
-        }
-      }
-    }
-    return result;
-  }
+  //  */
+  // getL() {
+  //   const result = new Matrix( this.m, this.n );
+  //   for ( let i = 0; i < this.m; i++ ) {
+  //     for ( let j = 0; j < this.n; j++ ) {
+  //       if ( i > j ) {
+  //         result.entries[ result.index( i, j ) ] = this.LU[ this.matrix.index( i, j ) ];
+  //       }
+  //       else if ( i === j ) {
+  //         result.entries[ result.index( i, j ) ] = 1.0;
+  //       }
+  //       else {
+  //         result.entries[ result.index( i, j ) ] = 0.0;
+  //       }
+  //     }
+  //   }
+  //   return result;
+  // }
+  //
+  // /**
+  //  * @public
+  //  *
+  //  * @returns {Matrix}
+  //  */
+  // getU() {
+  //   const result = new Matrix( this.n, this.n );
+  //   for ( let i = 0; i < this.n; i++ ) {
+  //     for ( let j = 0; j < this.n; j++ ) {
+  //       if ( i <= j ) {
+  //         result.entries[ result.index( i, j ) ] = this.LU[ this.matrix.index( i, j ) ];
+  //       }
+  //       else {
+  //         result.entries[ result.index( i, j ) ] = 0.0;
+  //       }
+  //     }
+  //   }
+  //   return result;
+  // }
 
   /**
    * @public
@@ -178,16 +188,16 @@
    *
    * @returns {number}
    */
-  det() {
-    if ( this.m !== this.n ) {
-      throw new Error( 'Matrix must be square.' );
-    }
-    let d = this.pivsign;
-    for ( let j = 0; j < this.n; j++ ) {
-      d *= this.LU[ this.matrix.index( j, j ) ];
-    }
-    return d;
-  }
+  // det() {
+  //   if ( this.m !== this.n ) {
+  //     throw new Error( 'Matrix must be square.' );
+  //   }
+  //   let d = this.pivsign;
+  //   for ( let j = 0; j < this.n; j++ ) {
+  //     d *= this.LU[ this.matrix.index( j, j ) ];
+  //   }
+  //   return d;
+  // }
 
   /**
    * @public
@@ -209,12 +219,14 @@
     // Copy right hand side with pivoting
     const nx = matrix.getColumnDimension();
     const Xmat = matrix.getArrayRowMatrix( this.piv, 0, nx - 1 );
+    const entries = [];
+    Xmat.entries.forEach(e => entries.push(new BigNumber(e)));
 
     // Solve L*Y = B(piv,:)
     for ( k = 0; k < this.n; k++ ) {
       for ( i = k + 1; i < this.n; i++ ) {
         for ( j = 0; j < nx; j++ ) {
-          Xmat.entries[ Xmat.index( i, j ) ] -= Xmat.entries[ Xmat.index( k, j ) ] * this.LU[ this.matrix.index( i, k ) ];
+          entries[ Xmat.index( i, j ) ] = entries[ Xmat.index( i, j ) ].minus(entries[ Xmat.index( k, j ) ].times(this.LU[ this.matrix.index( i, k ) ]));
         }
       }
     }
@@ -222,14 +234,17 @@
     // Solve U*X = Y;
     for ( k = this.n - 1; k >= 0; k-- ) {
       for ( j = 0; j < nx; j++ ) {
-        Xmat.entries[ Xmat.index( k, j ) ] /= this.LU[ this.matrix.index( k, k ) ];
+        entries[ Xmat.index( k, j ) ] = entries[ Xmat.index( k, j ) ].dividedBy(this.LU[ this.matrix.index( k, k ) ]);
       }
       for ( i = 0; i < k; i++ ) {
         for ( j = 0; j < nx; j++ ) {
-          Xmat.entries[ Xmat.index( i, j ) ] -= Xmat.entries[ Xmat.index( k, j ) ] * this.LU[ this.matrix.index( i, k ) ];
+          entries[ Xmat.index( i, j ) ] = entries[ Xmat.index( i, j ) ].minus(entries[ Xmat.index( k, j ) ].times(this.LU[ this.matrix.index( i, k ) ]));
         }
       }
     }
+
+    const efloats = entries.map(e=>e.toNumber());
+    Xmat.entries = efloats;
     return Xmat;
   }
 }
```

</details>

We can turn up the precision to get a more accurate result:

### 15 decimal places:

```
dim: 6x1
0.095781237 
0 
99999.99999990422 
99999.9999999042 
100000.00023704463 
100000 
```

### 20 decimal places

```
dim: 6x1
6.1854304e-7 
0 
100000 
100000 
100000.00000000368 
100000 
```


### 30 decimal places

```
dim: 6x1
-2.5093793e-17 
0 
100000 
100000 
100000 
100000 
```

### 100 decimal places

```
dim: 6x1
1.01516598e-86 
0 
100000 
100000 
100000 
100000 
```

### 1000 decimal places

```
dim: 6x1
0 
0 
100000 
100000 
100000 
100000 
```

**Comment by samreid** [comment]
By the way, here is the test harness I have been using:

```js
const stringToMatrix = string => {
  const data = string.split( '\n' ).map( row => row.trim().split( ' ' ) );
  const m = data.length;
  const n = data[ 0 ].length;
  const matrix = new Matrix( m, n );
  for ( let i = 0; i < m; i++ ) {
    for ( let k = 0; k < n; k++ ) {
      matrix.set( i, k, parseFloat( data[ i ][ k ] ) );
    }
  }
  return matrix;
};
const matrix = stringToMatrix( `0 1 0 0 0 0 
0 0 4831038889.801163 -4830038889.801163 0 -1000000 
0 0 -4830038889.801163 4830038889.901163 -0.1 0 
0 0 0 -0.1 0.1 0 
-1 0 -1000000 0 0 1000000 
0 -1 0 0 0 1` );

console.log( matrix.toString() );

const b = stringToMatrix( `0 
0 
0 
0 
0 
100000 ` );
console.log( b.toString() );

const x = matrix.solve( b );
console.log( x.toString() );
```

**Comment by samreid** [comment]
@jonathanolson and I discussed the proposed approach, and agreed it seems reasonable to move forward with decimal.js in sherpa, and adding HighPrecisionLUDecomposition in dot, which would be a copy-pasted version of LUDecomposition (the parts needed for a solve), so both can run at full speed.  We also tested and saw all matrices to be square, so we don't think we need to do QR (but if a non-square shows up, we will have to do QR).  HighPrecisionLUDecomposition will use decimal.js, but only sims that need HighPrecisionLUDecomposition will have to include decimal.js since it is 40KB+ minified.

I am hopeful it will solve this corner case, and it remains to be seen if it may solve other corner cases as well.

**Comment by samreid** [comment]
I have a patch with decimal.js working nicely, but in testing before I committed, I observed that even with 200 decimal places, you can easily create a circuit that has a bad solution:

![image](https://user-images.githubusercontent.com/679486/129409417-676a0abb-94c1-45f3-8058-437f19639964.png)

Not sure if I should commit the proposal as a partial solution, or whether we need to find another linear algebra solution.

**Comment by samreid** [comment]
I pushed what @jonathanolson and I discussed even though I can see it still fails on some easy cases.  Maybe there is still a bug or imprecise part in LUDecompositionDecimal, or maybe we need a new algorithm.

**Comment by samreid** [comment]
Here's a case that's still failing:

```
Debugging circuit: resistors:node1 -> node2 @ 2.0915055657130383e-10 Ohms,node2 -> node3 @ 10 Ohms,node4 -> node1 @ 0.000001 Ohms, batteries: node0 -> node4 @ 100000 Volts, currentSources: 
    equations:
    V0=0
1000000*V1+-1000000*V4+-4781244747.2932205*V2+4781244747.2932205*V1=0
4781244747.2932205*V2+-4781244747.2932205*V1+-0.1*V3+0.1*V2=0
0.1*V3+-0.1*V2=0
-I0_4+-1000000*V1+1000000*V4=0
-V0+V4=100000
    
    A=
dim: 6x6
0 1 0 0 0 0 
0 0 4782244747.2932205 -4781244747.2932205 0 -1000000 
0 0 -4781244747.2932205 4781244747.393221 -0.1 0 
0 0 0 -0.1 0.1 0 
-1 0 -1000000 0 0 1000000 
0 -1 0 0 0 1 

    
    z=
dim: 6x1
0 
0 
0 
0 
0 
100000 

    
    unknowns=
I0_4
V0
V1
V2
V3
V4
    
    x=
dim: 6x1
0.049999999999974995 
0 
99999.99999995 
99999.99999994999 
99999.99999994999 
100000 
```

**Comment by samreid** [comment]
I used math.js and lusolve, but it had the same bad error we started with:

<details>

```diff

Index: js/model/ModifiedNodalAnalysisCircuit.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/model/ModifiedNodalAnalysisCircuit.js b/js/model/ModifiedNodalAnalysisCircuit.js
--- a/js/model/ModifiedNodalAnalysisCircuit.js	(revision b2a91530f36d82cefb497d36afbb9ba66c7b12cb)
+++ b/js/model/ModifiedNodalAnalysisCircuit.js	(date 1628888249706)
@@ -330,43 +330,68 @@
     for ( let i = 0; i < equations.length; i++ ) {
       equations[ i ].stamp( i, A, z, getIndex );
     }
-    phet.log && phet.log( A.m, A.n );
-    phet.log && phet.log( `Debugging circuit: ${this.toString()}` );
-    phet.log && phet.log( equations.join( '\n' ) );
-    phet.log && phet.log( `A=\n${A.toString()}` );
-    phet.log && phet.log( `z=\n${z.toString()}` );
-    phet.log && phet.log( `unknowns=\n${unknowns.map( u => u.toTermName() ).join( '\n' )}` );
 
     // solve the linear matrix system for the unknowns
     let x;
-    try {
-      assert && assert( A.m === A.n, `the matrix should be square, instead it was ${A.m} x ${A.n}` );
-      x = new LUDecompositionDecimal( A, LUDecimal ).solve( z, LUDecimal );
+    // try {
+    assert && assert( A.m === A.n, `the matrix should be square, instead it was ${A.m} x ${A.n}` );
+    x = new LUDecompositionDecimal( A, LUDecimal ).solve( z, LUDecimal );
 
-      // const x1 = A.inverse().times( z );
-      // console.log( 'x' )
-      // console.log( x.toString() );
-      // console.log( 'x1' )
-      // console.log( x1.toString() );
-      //
-      // const a1 = new LUDecompositionDecimal( A, LUDecimal ).solve( Matrix.identity( A.m, A.m ), LUDecimal );
-      // const x2 = a1.times( z );
-      // console.log( 'x2' );
-      // console.log( x2.toString() );
-    }
-    catch( e ) {
+    // const x1 = A.inverse().times( z );
+    // console.log( 'x' )
+    // console.log( x.toString() );
+    // console.log( 'x1' )
+    // console.log( x1.toString() );
+    //
+    // const a1 = new LUDecompositionDecimal( A, LUDecimal ).solve( Matrix.identity( A.m, A.m ), LUDecimal );
+    // const x2 = a1.times( z );
+    // console.log( 'x2' );
+    // console.log( x2.toString() );
 
-      // Sometimes a fuzz test gives a deficient matrix rank.  It is a rare error and I haven't got one in the
-      // debugger yet to understand the cause.  Catch it and provide a solution of zeroes of the correct dimension
-      // See https://github.com/phetsims/circuit-construction-kit-dc/issues/113
-      x = new Matrix( A.n, 1 );
+    console.log( A.toString() );
+    const m = [];
 
-      // console.log( 'Rank deficient matrix!' )
+    for ( let i = 0; i < A.m; i++ ) {
+      const e = A.getMatrix( i, i, 0, A.n - 1 ).entries;
+      const x = [];
+      e.forEach( element => {
+        x.push( element );
+      } );
+      m.push( x );
     }
+    const zz = [];
+    z.entries.forEach( zzz => {
+      zz.push( zzz );
+    } );
+    const x3 = math.lusolve( m, zz);
+    console.log( 'x3' );
+    console.log( x3 );
+    // }
+    // catch( e ) {
+    //
+    //   // Sometimes a fuzz test gives a deficient matrix rank.  It is a rare error and I haven't got one in the
+    //   // debugger yet to understand the cause.  Catch it and provide a solution of zeroes of the correct dimension
+    //   // See https://github.com/phetsims/circuit-construction-kit-dc/issues/113
+    //   x = new Matrix( A.n, 1 );
+    //
+    //   // console.log( 'Rank deficient matrix!' )
+    // }
 
     // The matrix should be square since it is an exact analytical solution, see https://github.com/phetsims/circuit-construction-kit-dc/issues/96
     assert && assert( A.m === A.n, 'Matrix should be square' );
-    phet.log && phet.log( `x=\n${x.toString()}` );
+
+    console.log( `Debugging circuit: ${this.toString()}
+    equations:
+    ${equations.join( '\n' )}
+    
+    A=\n${A.toString()}
+    
+    z=\n${z.toString()}
+    
+    unknowns=\n${unknowns.map( u => u.toTermName() ).join( '\n' )}
+    
+    x=\n${x.toString()}
+    ` );
 
     const voltageMap = {};
     for ( let i = 0; i < unknownVoltages.length; i++ ) {


```

</details>

![image](https://user-images.githubusercontent.com/679486/129417894-9c9dd30c-d405-41b9-8cc8-b3f6b11e2764.png)

**Comment by samreid** [comment]
Here's the matrix problem:

```js
a = 1000000
b = 4781244747.2932205
V0 = 0
a*V1 -a*V4 -b*V2 + b*V1 = 0
b*V2 -b*V1 -0.1*V3 + 0.1*V2 = 0
0.1*V3 -0.1*V2 = 0
-I0_4 -a*V1 + a*V4 = 0
-V0 + V4 = 100000
```

solving through substitution (by hand) gives I0_4 = 0, but all the numerical methods put it around +/-0.01 or so.

**Comment by samreid** [comment]
The other part that is unsolved is that the circuit mostly has the correct solution, but as the wire is resized it goes through spots of bad solution:

![Kapture 2021-08-13 at 19 09 17](https://user-images.githubusercontent.com/679486/129430087-e1f98fc5-a7e5-4c16-a944-6be36e86529e.gif)

**Comment by samreid** [comment]
Anonymizing the values, here are snapshots of 2 pixels apart (bad on the left, good on the right):

![image](https://user-images.githubusercontent.com/679486/129430425-0aa8c602-ff52-4888-9c35-a9b04dda189a.png)

The only difference is in the wire resistance.

**Comment by samreid** [comment]
@jonathanolson and I met about this today and confirmed the matrices are ill-conditioned.  We used Mathematica and singularValueDecomposition[2] // MatrixForm and saw that it was typically 1E10 (high value) vs 0.01 (low value).  @jonathanolson identified a spot where an addition is done while creating the matrix, where there is an opportunity for floating point error.  We hypothesize that using arbitrary precision for that addition may stabilize the solution (even though the matrix remains ill-conditioned).  I'll test that next.

**Comment by samreid** [comment]
I added the higher precision for term addition, and it seems to stabilize the problem.  Even though the matrix is ill-conditioned, we arrive at a stable, more accurate result.  This is great news!  Testing with 200 points of precision in CCK-AC yields performance problems, so I think the next step would be to tune the precision/performance balance.

**Comment by samreid** [comment]
I defaulted the precision to 32 places, and added a query parameter "?precision" in case we want to tune it further.

**Comment by samreid** [comment]
I think this will be ready for testing in the next round.  On hold until other issues are fixed.

**Comment by samreid** [comment]
While working on, https://github.com/phetsims/circuit-construction-kit-common/issues/678, I observed this:

![Kapture 2021-08-18 at 13 45 08](https://user-images.githubusercontent.com/679486/129962336-fe034c03-29cd-4471-8bfa-fb836a6bb431.gif)

**Comment by samreid** [comment]
I requested QA testing for this issue.

**Comment by samreid** [comment]
In today's design meeting, I described the solution that runs the high precision solver 10 times before turning to the high performance solver.  Also, we saw a dangling circuit element that led to buggy behavior.  I described the ?precision=... query parameter to QA so they can see if turning up the precision can avoid errors like that.

**Comment by Nancy-Salpepi** [comment]
Using version 1.0.0-dev.20 on MacBook Air + chrome, while in the process of creating a circuit, the electrons moved, the power source went on fire and the bulb lit (even though it was still an incomplete circuit).

<img width="587" alt="Screen Shot 2021-09-07 at 3 14 27 PM" src="https://user-images.githubusercontent.com/87318828/132406723-69ddf2e4-af20-4132-9b28-59a4a48d6492.png">


Steps to reproduce the above: 
1. Place bulb in play area followed by an AC voltage source
2. Connect a wire to the bulb and then to the AC voltage
3. Connect a wire on the right side of the bulb
4. Connect a wire on the left side of the bulb (where the first wire was placed)
5. Wiggle the circuit around

![bulblit2](https://user-images.githubusercontent.com/87318828/132407107-36b224b5-b046-41bd-8bc9-0193f6570c49.gif)

**Comment by samreid** [comment]
Thanks for the great instructions @Nancy-Salpepi, I can easily produce this problem.  I tried with ?precision=1000 and still see the problem.  I tried ?capacitorResistance=1E-4 and still see the problem.  Maybe the solution is to eliminate "dangling" circuit elements from the matrix.  If a circuit element is not a member of a loop, then it doesn't contribute equations to the matrix, and doesn't get values from the matrix.

**Comment by samreid** [comment]
In the commit, I made it so that only circuit elements in a loop with a voltage source contribute to the circuit solution.  Any circuit element not in a loop with a voltage source will have its current zeroed out.  This was a complicated depth-first-search, and has not been well-vetted.  Open switches are not traversed.  Capacitors and inductors contribute a voltage if their companion model provides a voltage.

To test this issue, we should check that the bugs above no longer occur, but also that we still get the expected results in other scenarios.

I noticed that disconnecting a charged capacitor makes it look uncharged, but it still seems charged.  I'll open a side issue for that one.

**Comment by Nancy-Salpepi** [comment]
@sam reid
@stemilymill and I tried several several times to replicate the issues seen in https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-906668899 and https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-914612755 but could not. 

Can you elaborate on how you would like me to test: "but also that we still get the expected results in other scenarios?"

**Comment by samreid** [comment]
> Can you elaborate on how you would like me to test: "but also that we still get the expected results in other scenarios?"

Basically, testing other known scenarios, like:
* 2 resistors in series should add
* 2 resistors in parallel should have a harmonic sum like https://www.electronics-tutorials.ws/resistor/res_4.html
* battery resistor capacitor should have exponential decay with a time constant RC https://www.electronics-tutorials.ws/rc/rc_1.html
* switch behaves as expected
* etc

**Comment by samreid** [comment]
At today's design meeting, we agreed to hold off of testing until https://github.com/phetsims/circuit-construction-kit-common/issues/738 is complete.

**Comment by samreid** [comment]
#738 has a proposed solution, so this should be ready for testing in https://github.com/phetsims/circuit-construction-kit-common/issues/727

**Comment by Nancy-Salpepi** [comment]
When I rebuilt the first circuit (with no dangling wires)in https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-906668899 the circuit functioned correctly when the switch was open. There was no current reading and the battery voltage was additive. 

<img width="1177" alt="Screen Shot 2021-09-15 at 12 22 36 PM" src="https://user-images.githubusercontent.com/87318828/133471723-68be23b7-540f-44fd-bc73-a1a1559c417d.png">


When the switch was closed, there was electron bunching and there was current on one side of the circuit but not the other. I took the following screenshots several seconds apart to illustrate. 


<img width="985" alt="Screen Shot 2021-09-15 at 12 03 44 PM" src="https://user-images.githubusercontent.com/87318828/133470449-21904039-1bef-40aa-b981-7d0e6b7df0bb.png">
 

<img width="985" alt="Screen Shot 2021-09-15 at 12 04 18 PM" src="https://user-images.githubusercontent.com/87318828/133470479-2ee3053a-a13c-474a-b6f4-4877cdb4de5d.png">

<img width="1017" alt="Screen Shot 2021-09-15 at 12 05 30 PM" src="https://user-images.githubusercontent.com/87318828/133470699-56a1b43a-3284-485f-af10-3723f9d644f9.png">

**Comment by stemilymill** [comment]
I had similar outcomes to Nancy. Here are videos to demonstrate.


https://user-images.githubusercontent.com/85511101/133475539-f6106bab-b461-49af-bc09-2d44f8982903.mp4


Moving a dangling wire seemingly "resets" electrons

https://user-images.githubusercontent.com/85511101/133475550-66b6e5dd-d7ed-49b2-95b3-3d4f475f5b43.mp4



https://user-images.githubusercontent.com/85511101/133475564-6cfd9be1-895a-4f8b-a6a6-4be6f2696998.mp4


Opening and closing the switch

https://user-images.githubusercontent.com/85511101/133475578-faf18e8e-de9e-4c3c-96ce-771421a2f418.mp4

**Comment by samreid** [comment]
I build what I think is an equivalent circuit and did not see any current flowing:

![image](https://user-images.githubusercontent.com/679486/133478580-0f3d1580-4d6f-4ac8-adb1-f5a3f0cbc4d9.png)

Can you help track down the minimum number of steps to reproduce that problem?  Am I testing it wrong, or does it occur differently on different platforms?

**Comment by stemilymill** [comment]
This was win10 chrome, it might only occur sometimes. We can keep testing if needed, while keeping track of steps.

https://user-images.githubusercontent.com/85511101/133479509-a4f3f502-d07d-48de-8afd-9b8514f9cd5e.mp4

**Comment by Nancy-Salpepi** [comment]
@samreid the key for me to reproduce this issue was to wiggle bulb around. I hope these steps work for you!

I am on MacBook Air + chrome + 1.0.0-dev.25 
Here are the steps I was using:

1. Add 7 batteries from left to right across the bottom of the screen
2. In a clockwise fashion starting from the left battery, add a wire -- switch --wire --light bulb -- wire -- bill --wire to close the circuit.
3. Add ammeters on opposite sides of the circuit and voltage chart to opposite sides of batteries.
4. Check values box
5. Increase battery voltage for all batteries to 120V starting with leftmost battery
6. Decrease bulb resistance to zero
7.  Close switch
8. Wiggle bulb around until you see a current on either meter (I have been able to get it to appear on either one).

![wigglebulbcurrent](https://user-images.githubusercontent.com/87318828/133491028-9480448a-770e-4b40-a443-f7fd84e08a87.gif)

The electron bunching also occurs:

<img width="1001" alt="Screen Shot 2021-09-15 at 2 45 11 PM" src="https://user-images.githubusercontent.com/87318828/133491388-45bb77bf-9201-4b87-861f-3611eb22411c.png">

**Comment by stemilymill** [comment]
Win10 Chrome
I also recreated it at around the same time as Nancy, and made the circuit in similar order, but without changing the resistance of the lightbulb or checking the values box. I got current with and without a dangling circuit element.

Making the circuit:

https://user-images.githubusercontent.com/85511101/133492403-a1d0883b-75ea-4b05-9e61-198aeeae53a5.mp4

Add seven batteries left to right
Change voltage of batteries to 120 V left to right
Starting from left battery in clockwise order, add wire, then switch, wire, lightbulb, wire, dollar bill, wire connecting to right battery
Add voltmeter with sensors(?) on the very end of the wires connected to the batteries
Add ammeter to right bottom wire and left bottom wire
(if dangling) add wire to left battery connection
Close switch 
Wiggle lightbulb

With dangling wire:

https://user-images.githubusercontent.com/85511101/133492463-94deb605-c2e0-42c4-885f-e975ce6ef631.mp4

Without dangling wire:

https://user-images.githubusercontent.com/85511101/133493635-32e42c4f-df14-4b4e-bdd0-fd1602e56191.mp4

**Comment by kathy-phet** [comment]
Is this overall a lot harder to make "break" now??

**Comment by stemilymill** [comment]
Nancy and I were both able to quickly create this scenario with uneven current and the dollar bill, but it took multiple attempts to figure out a consistent process. However, the initial problem in this issue with current flowing with an open switch seems to be fixed but we would like to do more testing before saying for certain. We were sidetracked by the issues occurring in the closed circuit.

**Comment by Nancy-Salpepi** [comment]
@kathy-phet 
@stemilymill 

Using MacBook Air + chrome + 1.0.0-dev.25:

I created several different circuits the past hour--I was unable to produce a current when the switch was open or when there was an incomplete circuit no matter how much I wiggled the pieces around. When the circuit was closed, the voltage of the batteries in series was additive, the same for resistors. So, this part of the issue is either fixed or hard to reproduce. 

I continue to see problems with current/voltage when the circuit is closed and I wiggle things around. I am going to investigate that part further by comparing with dev.24 (or further back if needed). I can open a separate issue for that and link back to this one.

**Comment by kathy-phet** [comment]
Thanks, both! As we get to diminishing returns, and if this bug starts to seem like students would be going out of their way to discover it, then we should call it "solved" for now. I don't know that we will get to perfect behavior always. So please weigh in on that as you do more testing. If we are at 1/100,000 likelihood to be encountered by a user using the sim under normal conditions, we should probably turn the focus to other issues or publication.

**Comment by samreid** [comment]
Behavior looks good after the commits, but the new postprocessing steps introduce a lot of CPU time.  I need to investigate that before we start too much testing on a new version.

**Comment by samreid** [comment]
Ideas for moving forward:

* Depth first search should be vertex-based (instead of Path based as is committed)
* Search from participants first to have a consistent ground
* Avoid depth first search for anything already visited
* For a connected component, dfs should only run once.
* Open Switch needs to instigate a search from both vertices
* Run current conservation as a separate step (not part of depth first search).  Run the current conservation 3x, and make sure values settle to 0.  It should ignore 0 currents.

**Comment by samreid** [comment]
I rewrote the graph searches.  I used the stack-based search for loop detection, so that we could bail out early on loop detection.  However, this makes it difficult to detect if there was a voltage source in the stack.  I had some hesitation about requiring a voltage source in the stack anyways, so let's try this new rule of just detecting loops (or dangling branches) to determine what goes into the circuit solution.

**Comment by samreid** [comment]
On hold until https://github.com/phetsims/circuit-construction-kit-common/issues/739 has progress.

**Comment by samreid** [comment]
Summarizing the changes that require testing for this issue:
* The algorithms that detect loops and dangling wires have been rewritten to improve performance
* There is a new postprocessing algorithm that tries to pick up the slack in poorly conditioned solutions by helping to conserve current.  This should prevent electrons from bunching up in any circuit element.  This is a heuristic adjustment--not a guarantee, so we may still find cases where this is a problem.
* The "loop detection" algorithm used to require a voltage source in the loop, but now it includes any loop even with no voltage sources.  A reminder voltages sources are modeled in batteries, AC source, capacitor and inductor.

Requires testing:
* circuits with dangling branches
* circuits with disparate groupings (like separate circuits)
* circuits with loops that contain no voltage sources (see above for which circuit elements contain voltage sources).

**Comment by samreid** [comment]
@arouinfar and I were able to reproduce the problem.  It was very sensitive to a specific position/resistance of the wires.  It doesn't seem to be connected to the battery "shorting out" mode.  I'll look into it further, perhaps commenting out some of the new graph postprocessing steps to see if they are correlated.  But this is not reliable to reproduce every time, so may be difficult to know when there is a problem vs not.

**Comment by samreid** [comment]
I'll track the above problems in https://github.com/phetsims/circuit-construction-kit-common/issues/744

**Comment by stemilymill** [comment]
I was testing https://github.com/phetsims/circuit-construction-kit-common/issues/737 and repeatedly had current flowing across an open switch, but only in the switch itself. 
Here is a video taken with no added queries on win10 chrome.

https://user-images.githubusercontent.com/85511101/135327638-1175a8be-b84c-4c7f-b04d-316eb4e197ff.mp4

**Comment by brooklynlash** [comment]
I also saw this on Chromebook, with a simple circuit. You have to close the circuit, let it run a bit until an electron reaches the start of the switch, and then open it to easily replicate what we saw above.

**Comment by Nancy-Salpepi** [comment]
I thought this might be helpful in trying to determine why the switch has current (https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-930439605):
If you add a dangling wire to one of the switch nodes, there is no current reading for the switch.

<img width="634" alt="Screen Shot 2021-10-01 at 12 45 29 PM" src="https://user-images.githubusercontent.com/87318828/135657602-0ebaca7a-d23e-4abf-80f6-755038ac743a.png">

Steps to reproduce:
1. Create the circuit on the left starting from the battery and moving in a clockwise motion. 
2. Add the ammeter to the switch.
3. Close the switch and then open again--current reading for the switch says .90A
4. Add a dangling wire to one end of the switch
5. Close the switch and then open it. The current reading should be 0A

**Comment by Nancy-Salpepi** [comment]
For:
**Requires Testing**
- **circuits with dangling branches**

I created several different circuits and was unable to produce a current when the switch was open (except in the switch itself) or when a dangling wire was attached.  Adjusting the length of the wires did not produce a current in these cases.

**Comment by samreid** [comment]
The main changes for this issue were (a) the part that computes voltages for dangling branches was rewritten, (b) loops now do not require voltage sources and there is a new part that tries to correct numerical errors in the current conservation.

I think https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-930439605 is caused by the https://github.com/phetsims/circuit-construction-kit-common/issues/744

So the main things to test for #693 are just testing a few usual circuits, some with dangling branches.  Some circuits where there are loops that don’t connect to each other.

**Comment by KatieWoe** [comment]
A user reported this again, and it was reproduced in the published version of the sim on 5/12/22.
This was the set up:
<img width="1108" alt="Screen Shot 2022-05-11 at 12 11 30 PM" src="https://user-images.githubusercontent.com/41024075/168142816-2f05d99e-bc5f-4c93-ba89-0526c7c8fe4c.png">
Cut this point:
<img width="1154" alt="Screen Shot 2022-05-11 at 12 11 36 PM" src="https://user-images.githubusercontent.com/41024075/168142864-cd9644a3-03fb-44e3-9693-cf8776f5ade9.png">
This was the result:
<img width="1139" alt="Screen Shot 2022-05-11 at 12 11 42 PM" src="https://user-images.githubusercontent.com/41024075/168142896-c7b35505-4551-41cc-a26d-399fe066abd0.png">

When replicating, the results didn't seem so dramatic, but the lightbulb near the eraser still lit when it shouldn't. I also saw bunching electrons and more dramatic behavior while trying to build the circuit.

Result when recreating:
<img width="955" alt="happened" src="https://user-images.githubusercontent.com/41024075/168144089-f4fc3efa-f764-4c37-90ad-ebb54446fc7d.png">

**Comment by samreid** [comment]
I feel like https://github.com/phetsims/circuit-construction-kit-common/issues/228#issuecomment-1204572936 will be the most efficient way out of this problem.  @arouinar or @kathy-phet can you please help prioritize?

**Comment by oliver-phet** [comment]
Just documenting another user report of this:
![Construire un circuit _ CC screenshot](https://user-images.githubusercontent.com/5090275/196540788-c1dbc211-b2a2-4f66-adfe-dda85dbf4bfa.png)
Looks like electron bunching.

**Comment by samreid** [comment]
For me, this is another vote for https://github.com/phetsims/circuit-construction-kit-common/issues/228#issuecomment-1204572936

**Comment by arouinfar** [comment]
@samreid I am going to defer this issue. I agree the solution is likely #228.

While this is indeed a serious bug, it is infrequent, and will largely be addressed by publishing the 1.3 version of CCK: DC which has model improvements developed during the initial publication of CCK: AC.

The user reports in this issue all come from the 1.2 version of CCK: DC, which predates the initial release of CCK: AC. We are in the process of publishing the 1.3 version of CCK: DC and it hasn't reoccurred during QA testing (yet, at least).

**Comment by Nancy-Salpepi** [comment]
I tried for a while to reproduce this issue in 1.3.0-rc.1, but could not. I was able to easily reproduce this case https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-1282982568 in the published version (to make circuit I started at the bulb and moved in a clockwise direction).
